spacenav: init package and module

lukego opened the pull request:
###### Motivation for this change

This PR adds hardware support for the 3Dconnexion Spacemouse (https://www.youtube.com/watch?v=4WBTJhXpnsc) to nixpkgs in general and Blender in particular.

Specifically,

- Package the [Spacenav](http://spacenav.sourceforge.net/) software packages `spacenavd`, `libspnav`, `spnavcfg`.
- Add option `spacenavSupport` to Blender (enabled by default);
- Patch Blender to recognize the device ID of the hardware unit that I bought in 2020 (have to contact Blender upstream about that patch separately.)

This works great for me! However I have to start `spacenavd` manually as root. (I tried writing a NixOS module to operate that but my first attempt hasn't worked and I'm a bit of a systemd n00b.)

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

lukego committed:
spacenav: Added spacenavd, libspnav, spnavcfg packages

Supports 3D input devices such as 3Dconnexion Spacemouse.

lukego committed:
blender: Add spacenavSupport option, enabled by default

Builds Blender with support for 3D input devices such as 3Dconexxsion
Spacemouse.

Patched Blender to recognize the device-ID of the wireless Spacemouse
that I bought in 2020. Has to be upstreamed to Blender separately.

Enabled by default even though the feature is obscure because enabling
has low impact (small dependency) while enabling manually has larger
impact (have to compile Blender which takes a long time.)

prusnak commented on pkgs/misc/spacenav/libspnav.nix:
```diff
@@ -0,0 +1,13 @@
+stdenv.mkDerivation {
+  pname = "libspnav";
+  version = "0.2.3";
+  src = fetchurl {
+    url = "https://github.com/FreeSpacenav/libspnav/releases/download/libspnav-0.2.3/libspnav-0.2.3.tar.gz";
+    sha256 = "14qzbzpfdb0dfscj4n0g8h8n71fcmh0ix2c7nhldlpbagyxxgr3s";
+  };
```
```suggestion
  src = fetchFromGitHub {
    owner = "FreeSpacenav";
    repo = pname;
    rev = "v${version}";
    sha256 = "xxx";
  };
```

prusnak commented on pkgs/misc/spacenav/spacenavd.nix:
```diff
@@ -0,0 +1,13 @@
+stdenv.mkDerivation {
+  pname = "spacenavd";
+  version = "0.7.1";
+  src = fetchurl {
+    url = "https://github.com/FreeSpacenav/spacenavd/releases/download/v0.7.1/spacenavd-0.7.1.tar.gz";
+    sha256 = "0awxa90rqmpvwa04fy1wz0hq0cq4c579h5hgmjdwgavxhjdy2rlj";
+  };
```
```suggestion
  src = fetchFromGitHub {
    owner = "FreeSpacenav";
    repo = pname;
    rev = "v${version}";
    sha256 = "xxx";
```

prusnak commented on pkgs/misc/spacenav/spnavcfg.nix:
```diff
@@ -0,0 +1,16 @@
+stdenv.mkDerivation {
+  pname = "spnavcfg";
+  version = "0.3";
+  src = fetchurl {
+    url = "https://github.com/FreeSpacenav/spnavcfg/releases/download/spnavcfg-0.3/spnavcfg-0.3.tar.gz";
+    sha256 = "1msbxb6i01irrc5b5dvbxirwzf415y9cbdh8pz046hmx9s6hp5ac";
+  };
```
```suggestion
  src = fetchFromGitHub {
    owner = "FreeSpacenav";
    repo = pname;
    rev = "v${version}";
    sha256 = "xxx";
```

lukego committed:
Update pkgs/misc/spacenav/spnavcfg.nix

Co-authored-by: Pavol Rusnak <pavol@rusnak.io>

lukego committed:
Update pkgs/misc/spacenav/spacenavd.nix

Co-authored-by: Pavol Rusnak <pavol@rusnak.io>

lukego committed:
Update pkgs/misc/spacenav/libspnav.nix

Co-authored-by: Pavol Rusnak <pavol@rusnak.io>

lukego commented:
Oh dear. I applied those suggestions that Github offered me but it seems they were only drafts and not actually working. Sorry, haven't seen this workflow on Github before with being offered changes to adopt, didn't have a workflow for testing/evaluating them beyond a quick eyeball.

Closing this PR because the review process is a bit overwhelming with all the advanced Github features and bot reviews :)

KoviRobi commented:
I just got a SpacePilot Pro, and am interested in helping you test (and could implement a NixOS module too), we'll see if I can find the time to do so :)

Commit:
spacenav: Move misc/spacenav to tools/inputmethods/

Commit:
spacenav: Fix compilation of tools/inputmethods/spacenav/*

Commit:
openscad: Add 3D mouse support

KoviRobi commented:
I don't know if there is a better way to suggest changes, but I have a couple of commits on top of yours, at https://github.com/KoviRobi/nixpkgs/commits/spacenav-nixpkgs. I reopened this, as this work is still based you yours, but I can close this and create a new PR if you want.

If you want to fold my changes into yours, you should be able to do (I think)
```
git remote add kovirobi-spacenav https://github.com/KoviRobi/nixpkgs.git
git fetch kovirobi-spacenav
git rebase kovirobi-spacenav/spacenav-nixpkgs
```

Commit:
nixos/spacenavd: Add spacenavd service

This runs the driver for 3D mice such as Space Navigator, Space Pilot or
Space mouse.

KoviRobi commented:
I have also added a service, which is currently run as root :( but I couldn't get it working otherwise.

lukego commented:
That's awesome @KoviRobi! I've pulled your commits onto this branch. If my Git-fu is lacking in some respect feel free to open a new PR from a branch of your own and we can close this one.

7c6f434c reviewed (approved):
Code looks good to me

flokli commented on pkgs/applications/misc/blender/3dconnexion-spacemouse-wireless.patch:
```diff
@@ -0,0 +1,13 @@
+     case 0x256F:  // 3Dconnexion
+       switch (product_id) {
+         case 0xC62E:  // plugged in
+-        case 0xC62F:  // wireless
```
Is this somewhere upstream in blender, and could we `fetchpatch` from there?

flokli commented on nixos/modules/services/hardware/spacenavd.nix:
```diff
@@ -0,0 +1,30 @@
+      wantedBy = [ "multi-user.target" ];
+      partOf = [ "multi-user.target" ];
+      serviceConfig = {
+        ExecStart = "${pkgs.spacenavd}/bin/spacenavd -d";
```
Last time I played with this, I opened 3 issues:

 - https://github.com/FreeSpacenav/spacenavd/issues/8
   This allows specifying a custom config file. We currently don't seem to configure anything?
 - https://github.com/FreeSpacenav/spacenavd/issues/9
  This introduces `-l syslog`, which we should probably set, to properly capture log output
 - https://github.com/FreeSpacenav/spacenavd/issues/10
  The `spnav.sock` socket currently is hardcoded, and I'm not sure about its permissions. Did you verify this works?

KoviRobi replied:
I did manage to get spacenav working with this, but it's worth pointing out that spacenav joystick support is in Linux out of the box, and you can disable the annoying mouse movements with this:
```
  services.xserver.inputClassSections = [
    ''
      Identifier "3Dconnexion  SpacePilot PRO "
      MatchProduct "3Dconnexion  SpacePilot PRO "
      Driver "evdev"
      Option "Ignore" "on"
    '' ];

```
But blender doesn't support mouse navigation (openscad does).

- The config file is something which we could potentially capture with nix, but given it's got a gui configurator and it's very floating-point not easy to calculate you just have to try and see what works, we do definitely want to support that.
- Good idea
- I did verify, it works (when it's launched as root, I couldn't make it work non root though :( )

flokli replied:
Yeah, this could use some documentation, maybe the inputClassSection even a NixOS option.

Can you add some module documentation, like it's done with the nextcloud module?

> The config file is something which we could potentially capture with nix, but given it's got a gui configurator and it's very floating-point not easy to calculate you just have to try and see what works, we do definitely want to support that.

So you want to give this some cooking first?

> I did verify, it works (when it's launched as root, I couldn't make it work non root though :( )

GUI applications need to be launched as root? Or the spacenav daemon?

KoviRobi replied:
The spacenav daemon needs to be root -- though you do need write permissions to the /etc/spnav.conf, which currently I think requires a chown.

Also, because my brain is a vegetable today, I forgot your first point, the OP did say they tried to get their patch accepted upstream, not sure if they have heard anything back from them.

I'll have a think about the config file, what I meant though is that it might not be applicable to be captured by nix, as the user might want to change it too often/faster than recompiling the system/with the gui rather than just picking values.

I'll try and remember to look at this in a couple of days when I am more functional, thanks for taking a look at this PR though!

flokli replied:
Having to no clear separation between user-wanted config and "state/calibration data" would indeed be a problem - let's iterate on that at a later point :-)

prusnak commented:
These three issues are duplicates of each other:
* #75866
* #94234
* #108838

leenaars commented:
And there is a version bump for spnavcfg too:

```
  src = fetchFromGitHub {
    owner = "FreeSpacenav";
    repo = pname;
    rev = "v${version}";
    hash = "sha256-QCHAMua4Wtn+v2fw5VdA29yg+IXHjZzT0f2VoGObFaA=";
  };
```

Sohalt commented:
Can you also add the `spacenav-cube-example`? You can have a look at https://github.com/NixOS/nixpkgs/pull/75866 for the rationale and implementation.

flokli commented:
For posterity: This has been implemented in https://github.com/NixOS/nixpkgs/pull/118617.
